### ISSUE TITLE
Move IP handling to Packet Analysis architecture

### ISSUE THREAD

**Issue by timwoj** [zeek/zeek#1131]
This work is predicated on the work for #248 being merged, but we can start talking about what this actually means.

### Current setup in packet_analysis branch

Packet received
Packet calls packet analysis to do setup of the packet fields
PktSrc calls net_packet_dispatch
net_packet_dispatch calls Session::NextPacket
Session::NextPacket does analysis for IPv4/IPv6 and calls L3 analyzers

### Proposed setup moving IP to packet_analysis

Packet received
PktSrc calls net_packet_dispatch
net_packet_dispatch calls packet_analysis
packet_analysis does full header analysis including IPv4/IPv6
Packet analyzers call L3 analyzers (ARP-PA calls ARP-L3A, IPv4/IPV6-PA call other L3 analyzers or back into Sessions for the existing Conn code?)

### Thoughts

- There’s a lot of duplicated code between handling of v4 and v6. There should probably be some sort of common location for this.
- IP-in-IP encapsulation (see DoNextPacket/DoNextInnerPacket) can be just a call back into the same analyzer. This would strip that part of the header checking out of DoNextPacket. DoNextInnerPacket can probably be removed entirely at that point.
- GRE tunnels can be in a separate analyzer
- TCP/UDP/ICMP connection tracking can stay in Sessions for now (#1114 will eventually move them)

### Questions/Issues

- Session::NextPacket does all of the packet profiling and dumping before calling DoNextPacket
    - This could be moved to net_packet_dispatch before and after calling packet analysis
- Session keeps statistics for the number of packets received
    - This could move out of session and into Net in general
- Packet filtering and discarding happens before header processing. Should this be moved Into packet analysis?
- Fragments are handled in Session::DoNextPacket before the header processing. Should this be moved into the packet analysis?

**Comment by timwoj** [comment]
> Fragments are handled in Session::DoNextPacket before the header processing. Should this be moved into the packet analysis?

@rsmmr and I talked about this one for a minute. Fragment handling is a feature of IP and should probably move into the packet analyzer.

> Packet filtering and discarding happens before header processing. Should this be moved Into packet analysis?

We also discussed this. Filtering and discarding can be called by the IP analyzer once it has enough information to bail out of the analysis if need be.

**Comment by rsmmr** [comment]
> ### Proposed setup moving IP to packet_analysis

This looks good to me (and more natural than the current path).

> Packet analyzers call L3 analyzers (ARP-PA calls ARP-L3A, IPv4/IPV6-PA call other L3 analyzers or back into Sessions for the existing Conn code?)

I think I'd go back into Sessions at least for the 1st version of this, so that we don't change too much at once. Can change more later. 

> * There’s a lot of duplicated code between handling of v4 and v6. There should probably be some sort of common location for this.

Ack.

> * IP-in-IP encapsulation (see DoNextPacket/DoNextInnerPacket) can be just a call back into the same analyzer. This would strip that part of the header checking out of DoNextPacket. DoNextInnerPacket can probably be removed entirely at that point.

That would be nice indeed!

> * GRE tunnels can be in a separate analyzer

Yep.

> Questions/Issues

As we chatted about, I think the general guideline here can be:

1. if it's something directly part of understanding a protocol, it should go with analyzer (e.g., IP defragmentation).
2. If it's something that uses analyzer-specific information (e.g., IP addresses), but is otherwise independent of analyzing the protocol (packet filtering & discarding, statistics), we can keep it as a central global component that the analyzer calls into once it has the information (e.g., filtering: IP calls into filtering and aborts packet processing depending on result).

With (2), the affected global functionality will remain restricted to the particular protocols that implement the call-outs, but that seem's fine given it's protocol-specific information to begin with. In the future, we could further generally this so that, e.g., the packet filter can work on other information than IP addresses as well, and then feed that from more places; but that seems out of scope for now. Main goal for these things is to retain current functionality so that we don't break existing setups.

**Comment by timwoj** [comment]
Diving into this a little more deeply, `run_state::dispatch_packet()` might need to be broken up a little bit too. Right now I was hoping to call `sessions->NextPacket()` directly from the packet analysis, as that feels the most natural. The problem with that plan is that `PktSrc::Process()` depends on the packet analysis succeeding and setting the `layer2valid` variable in the packet to determine whether `dispatch_packet()` is called. Without this, network time is never advanced and events are never triggered, amongst other things. `PktSrc::Process()` also handles whether the time should be updated based on pseudo-time or the packet time.

I don't think any of the above is a big deal, it's just a matter of finding the right places to handle everything.

**Comment by timwoj** [comment]
There is a number of things that happen in between packet analysis finishing and the rest of the analysis starting. For example, `raw_packet` events being enqueued (at the start of NetSessions::NextPacket), anything involving checking for whether `Packet::Layer2Valid` returns true, or the pseudo-time handling in PktSrc. At what point do we consider packet analysis done so that these things can happen? `raw_packet` events feel like they should happen for every single packet, whether or not they’re going into the IP analysis at all. I could add a method to Packet called `SetLayer2Valid()` or some such that does some of this, but Packet feels like the wrong place to do it.

My concern here is that there’s a lot of things that happen in Session that plugin authors don’t need to concern themselves with. Moving these out of Session means that now plugin authors need to make sure to call them or something isn’t going to work correctly.

Along the lines of not wanting to add anything extra to `Packet`, I’m debating if moving the time-handling code out of `PktSrc` and into `RunState` is a good idea. This makes more sense now that we only allow a single packet source at a time, so functions like pseudo-time don’t necessarily need to be handled in the packet source code.

**Comment by timwoj** [comment]
>> IP-in-IP encapsulation (see DoNextPacket/DoNextInnerPacket) can be just a call back into the same analyzer. This would strip that part of the header checking out of DoNextPacket. DoNextInnerPacket can probably be removed entirely at that point.

> That would be nice indeed!

This is called directly by the existing tunnel plugins though. Would you agree to those just becoming packet analysis plugins? `DoNextInnerPacket` could just be marked deprecated with an indication that tunnel plugins should be packet analysis plugins now.

**Comment by J-Gras** [comment]
> There is a number of things that happen in between packet analysis finishing and the rest of the analysis starting. For example, `raw_packet` events being enqueued (at the start of NetSessions::NextPacket), anything involving checking for whether `Packet::Layer2Valid` returns true, or the pseudo-time handling in PktSrc.

Having a completely flexible stack of packet analyzers (including IP) raises the question when `raw_packet` should be called? I would keep it before IP and eventually rename the event. I think the calls to `Packet::Layer2Valid()` will go away once we rework the `l2_hdr` record.

> Along the lines of not wanting to add anything extra to `Packet`, I’m debating if moving the time-handling code out of `PktSrc` and into `RunState` is a good idea.

I like the idea. Currently zeek seems to ignore packets with invalid L2. What's the rationale behind this? As far as I can tell the timestamp is provided by the packet source. The proposed changes (moving packet analysis to `dispatch_packet()` could slightly change semantics here. Note that currently packet analysis also happens before network time is updated. I think that should be changed.

**Comment by rsmmr** [comment]
A bit hard to tell without really digging into this, but from a quick look, ,oving the timing stuff out of Packet sounds good to me. Looks like PktSrc::Process() could move out of the class altogether.

And agree that plugin authors should need to care as little as possible outside of their specific protocol. That doesn't mean that for some built-in protocols, we can't insert calls to replicate current functionality, like if we always did something just before IP, the IP analyzer can continue to trigger that before everything else.

Yes, Packet::Layer2Valid() should go away. Isn't that covered now simply by the processing pipeline proceeding? I.e., the fact the Ethernet passes data on to IP, means it's good.

**Comment by timwoj** [comment]
> Note that currently packet analysis also happens before network time is updated. I think that should be changed.

I've been thinking about this one too. If we got a packet, shouldn't that packet just update the time? Is there any reason to wait for packet analysis to finish first?

> Having a completely flexible stack of packet analyzers (including IP) raises the question when raw_packet should be called? I would keep it before IP and eventually rename the event.

The real question was where to actually do it. We don't want to have to make plugin authors call it for everything to work as expected. Maybe it just stays as part of the IP analyzer for now and we can move it somewhere better once that work is completed? To be honest, it feels like an event that should just happen whether or not analysis succeeds, so maybe it should just happen in `PktSrc::Process()` before the actual analysis starts.

> Yes, Packet::Layer2Valid() should go away. Isn't that covered now simply by the processing pipeline proceeding? I.e., the fact the Ethernet passes data on to IP, means it's good.

Very true. If packet analysis succeeds, then the layer 2 headers are valid and there's no need for a check anymore.

**Comment by J-Gras** [comment]
> Maybe it just stays as part of the IP analyzer for now and we can move it somewhere better once that work is completed?

That's what I had in mind.

> To be honest, it feels like an event that should just happen whether or not analysis succeeds, so maybe it should just happen in `PktSrc::Process()` before the actual analysis starts.

In that case we won't be able to build the [raw_pkt_hdr](https://docs.zeek.org/en/current/scripts/base/init-bare.zeek.html#type-raw_pkt_hdr). Each analyzer could generate an analyzer-specific event. The downside would be that we lose the context of the packet.

**Comment by timwoj** [comment]
Is there any reason to avoid calling directly into `NetSessions::DoNextPacket` from the IP analyzer? We already know that all of the analyzers handled by NetSessions are IP-based. I started adding a more-generic system, but it really ends up being the IP analyzer tied to the `sessions` object in one way or another.

**Comment by timwoj** [comment]
I’ve been working on this for a bit. Here’s an update on the design.

- The packet processing path ended up just like the suggested design in the first comment: PktSrc -> dispatch_packet -> packet analysis -> session analysis (if an IP packet)
- All of the time-handling moved out of PktSrc into run_state.
- All of the IP-specific code was moved into a new packet analyzer. This handles both IPv4 and IPv6 in a single analyzer, and so the other two analyzers were removed. This includes all of the IP-specific functionality such as packet filtering, discarding, and packet fragmentation.
- Packet profiling and dumping moved into the packet analysis manager.
- New GRE and IPTunnel packet analyzers were added to handle those parts of the analysis, since they are separate from the IP analyzer.

TODO/Questions:

- The `raw_packet` event now occurs after all of the analysis has finished in packet_mgr. It previously happened in between packet header analysis and the session analysis, since it relies on the packet header data being available to output. Does anything break with moving that later in the process?
- A key/store is being used in the Packet class to store data that needs to be passed between packet analyzers, such as the encapsulation data from the IP tunnel analyzer back to the IP analyzer. I have some concerns about the performance implications of using a `map<string,any>` for this. @j-gras, @rsmmr, and I discussed it a bit and there aren’t any great options. The performance implications of this need to be measured.
- The packet’s hdr_size is calculated one time, during the IP analyzer. I feel like this should be built up during the packet analysis chain, since the individual analyzers know how big their headers are and could easily just add to the value. This might be something to do in the other PR before it lands.
- The sessions code expects that the hdr_size value *doesn’t* include the size of the IP header, because it uses that to recalculate the capture length before anything else and because the raw packet header output relies on that to rebuild the IP header again for output. This doesn’t feel like it should be the case, especially if the above bullet gets implemented.
- In the existing code, the data pointer is updated to point at the IP header's payload after a fragmented packet is rebuilt so that the data pointer points at the full data for the session analyzer. This isn’t a big deal for when we pass a rebuilt packet directly to the session analyzer, since we pass the IP_Hdr object along and it just grabs the payload pointer before doing anything else and completely ignores the packet->data pointer. Is there ever a case where a fragmented packet being rebuilt doesn’t fall directly into the session analyzer? In that case, we need to update both the data pointer and the packet’s data pointer so that the later analyzers work correctly. Does this break anything?
- While the calls to dump packets moved into the packet analysis manager, the actual code still lives in Sessions. It should move.
- A number of public methods in Sessions related to packet handling will likely be removed. For sure going away are the two DoNextInnerPacket methods, since all of that code moved into the IPTunnel analyzer. The only places where I was able to determine these were called were from the existing tunnel analyzers, which are becoming packet analyzers anyways. The other method is NextPacket, which was previously the main starting point for session analysis. The IP analyzer calls DoNextPacket directly, and all of the code in NextPacket is obsolete now (in that it doesn’t work anyways). It was previously only called from dispatch_packet.

**Comment by timwoj** [comment]
Teredo, GTP, and Ayiya are a bit more difficult because they're tunneled over UDP. That means we would come out of packet analysis land and then discover those protocols, and want to go back into packet analysis. At least, until TCP and UDP are packet analyzers as well (#1114).

**Comment by rsmmr** [comment]
Going through the questions (I guess some are addressed/progressed already):

> * The `raw_packet` event now occurs after all of the analysis has finished in packet_mgr. It previously happened in between packet header analysis and the session analysis, since it relies on the packet header data being available to output. Does anything break with moving that later in the process?

Seems fine to me. I wouldn't expect important dependencies on ordering here. 

> * A key/store is being used in the Packet class to store data that needs to be passed between packet analyzers, such as the encapsulation data from the IP tunnel analyzer back to the IP analyzer. I have some concerns about the performance implications of using a `map<string,any>` for this. @J-Gras, @rsmmr, and I discussed it a bit and there aren’t any great options. The performance implications of this need to be measured.

XXX

> * The packet’s hdr_size is calculated one time, during the IP analyzer. I feel like this should be built up during the packet analysis chain, since the individual analyzers know how big their headers are and could easily just add to the value. This might be something to do in the other PR before it lands.

Ack, I like that better, too.

> * The sessions code expects that the hdr_size value _doesn’t_ include the size of the IP header, because it uses that to recalculate the capture length before anything else and because the raw packet header output relies on that to rebuild the IP header again for output. This doesn’t feel like it should be the case, especially if the above bullet gets implemented.

Agree, can just adapt the sessions code?

> Is there ever a case where a fragmented packet being rebuilt doesn’t fall directly into the session analyzer?

Not quite sure, but my guess is: no. Defragemention is really just preprocessing step before we can continue normally.

 > In that case, we need to update both the data pointer and the packet’s data pointer so that the later analyzers work correctly. Does this break anything?

I'm wondering if we should still do that even if technically we wouldn't need to, just to avoid confusion re/ inconsistent. Wouldn't expect anything to rely on this (and test-suite should catch if that's wrong ...)

> * While the calls to dump packets moved into the packet analysis manager, the actual code still lives in Sessions. It should move.

Ack.

> * A number of public methods in Sessions related to packet handling will likely be removed. 

Ack.  Will be fine generally overhaul this as desired, nothing should depend on internal structure.

**Comment by timwoj** [comment]
> No good idea here still how to do this. How generic does this need to be? Can we restrict the kinds of information that need be handed around to a small set of options, or do we essentially not know what future analyzers might want to pass around?

I think we generally don't know what data can be passed around. I did run some benchmarks on the IP branch as it is now. It's about 1.5% slower, but I can't say how much of that could be attributed to the use of a `map` as I'm currently doing.

> Ack, I like that better, too.

This one became https://github.com/zeek/zeek/issues/1186, which I'll probably fix before the IP work lands makes it to a PR.

> Agree, can just adapt the sessions code?

Yah, I'll have to double check that a change there doesn't cause any weird issues down the line as the header size changes.

> Ack. Will be fine generally overhaul this as desired, nothing should depend on internal structure.

`DoNextInnerPacket` has been removed. `NextPacket` basically just turned into a call to `packet_mgr->ProcessPkt()` and was marked as deprecated to be removed.
